--- /usr/lib/python2.7/site-packages/trac/db/mysql_backend.py.orig	2018-07-29 19:16:12.000000000 +0100
+++ /usr/lib/python2.7/site-packages/trac/db/mysql_backend.py	2018-12-21 17:22:37.923493039 +0000
@@ -17,7 +17,6 @@
 import os
 import re
 import sys
-import types
 
 from genshi.core import Markup
 
@@ -34,16 +33,19 @@
 
 _like_escape_re = re.compile(r'([/_%])')
 
+has_pymysql = False
+pymysql_version = None
+
 try:
-    import MySQLdb
-    import MySQLdb.cursors
+    import pymysql
 except ImportError:
-    MySQLdb = None
-    mysqldb_version = None
+    raise TracError(_("Failed to import pymysql"))
+    pass
 else:
-    mysqldb_version = get_pkginfo(MySQLdb).get('version', MySQLdb.__version__)
+    has_pymysql = True
+    pymysql_version = get_pkginfo(pymysql).get('version', pymysql.__version__)
 
-    class MySQLUnicodeCursor(MySQLdb.cursors.Cursor):
+    class MySQLUnicodeCursor(pymysql.cursors.Cursor):
         def _convert_row(self, row):
             return tuple(v.decode('utf-8') if isinstance(v, str) else v
                          for v in row)
@@ -62,6 +64,10 @@
             return [self._convert_row(row) for row in rows] \
                    if rows is not None else []
 
+    class MySQLSilentCursor(MySQLUnicodeCursor):
+        def _show_warnings(self, conn=None):
+            pass
+
 # Mapping from "abstract" SQL types to DB-specific types
 _type_map = {
     'int64': 'bigint',
@@ -95,14 +101,20 @@
         """Location of mysqldump for MySQL database backups""")
 
     def __init__(self):
-        self._mysql_version = None
+        if has_pymysql:
+            self._mysql_version = \
+                'server: (not-connected), client: "%s", thread-safe: %s' % \
+                (pymysql.get_client_info(), pymysql.thread_safe())
+        else:
+            self._mysql_version = None
+        self.error = None
 
     # ISystemInfoProvider methods
 
     def get_system_info(self):
-        if self.required:
+        if has_pymysql:
             yield 'MySQL', self._mysql_version
-            yield 'MySQLdb', mysqldb_version
+            yield pymysql.__name__, pymsql_version
 
     # IDatabaseConnector methods
 
@@ -115,13 +127,13 @@
         if not self.required:
             self._mysql_version = \
                 'server: "%s", client: "%s", thread-safe: %s' \
-                % (cnx.cnx.get_server_info(), MySQLdb.get_client_info(),
-                   MySQLdb.thread_safe())
+                % (cnx.cnx.get_server_info(), pymysql.get_client_info(),
+                   pymysql.thread_safe())
             self.required = True
         return cnx
 
     def get_exceptions(self):
-        return MySQLdb
+        return pymysql
 
     def init_db(self, path, schema=None, log=None, user=None, password=None,
                 host=None, port=None, params={}):
@@ -370,7 +382,7 @@
             password = ''
         if port is None:
             port = 3306
-        opts = {}
+        opts = {'charset': 'utf8'}
         for name, value in params.iteritems():
             key = name.encode('utf-8')
             if name == 'read_default_group':
@@ -381,22 +393,32 @@
                 opts[key] = value.encode(sys.getfilesystemencoding())
             elif name in ('compress', 'named_pipe'):
                 opts[key] = as_int(value, 0)
+            elif name == 'charset':
+                value = value.lower()
+                if value in ('utf8', 'utf8mb4'):
+                    opts[key] = value
+                else:
+                    self.log.warning("Invalid connection string parameter "
+                                     "'%s=%s'", name, value)
             else:
                 self.log.warning("Invalid connection string parameter '%s'",
                                  name)
-        cnx = MySQLdb.connect(db=path, user=user, passwd=password, host=host,
-                              port=port, charset='utf8', **opts)
-        self.schema = path
-        if hasattr(cnx, 'encoders'):
-            # 'encoders' undocumented but present since 1.2.1 (r422)
-            cnx.encoders[Markup] = cnx.encoders[types.UnicodeType]
+        cnx = pymysql.connect(db=path, user=user, passwd=password, host=host,
+                              port=port, **opts)
         cursor = cnx.cursor()
         cursor.execute("SHOW VARIABLES WHERE "
                        " variable_name='character_set_database'")
         self.charset = cursor.fetchone()[1]
-        if self.charset != 'utf8':
-            cnx.query("SET NAMES %s" % self.charset)
-            cnx.store_result()
+        cursor.close()
+        if self.charset != opts['charset']:
+            cnx.close()
+            opts['charset'] = self.charset
+            cnx = pymysql.connect(db=path, user=user, passwd=password,
+                                  host=host, port=port, **opts)
+        self.schema = path
+        if hasattr(cnx, 'encoders'):
+            # 'encoders' undocumented but present since 1.2.1 (r422)
+            cnx.encoders[Markup] = cnx.encoders[unicode]
         ConnectionWrapper.__init__(self, cnx, log)
         self._is_closed = False
 
@@ -407,14 +429,14 @@
         self.cnx.ping()
         try:
             self.cnx.rollback()
-        except MySQLdb.ProgrammingError:
+        except pymysql.ProgrammingError:
             self._is_closed = True
 
     def close(self):
         if not self._is_closed:
             try:
                 self.cnx.close()
-            except MySQLdb.ProgrammingError:
+            except pymysql.ProgrammingError:
                 pass # this error would mean it's already closed.  So, ignore
             self._is_closed = True
 
@@ -429,7 +451,7 @@
         return 'concat(%s)' % ', '.join(args)
 
     def drop_column(self, table, column):
-        cursor = MySQLdb.cursors.Cursor(self.cnx)
+        cursor = pymysql.cursors.Cursor(self.cnx)
         if column in self.get_column_names(table):
             quoted_table = self.quote(table)
             cursor.execute("SHOW INDEX FROM %s" % quoted_table)
@@ -451,8 +473,7 @@
                            (quoted_table, self.quote(column)))
 
     def drop_table(self, table):
-        cursor = MySQLdb.cursors.Cursor(self.cnx)
-        cursor._defer_warnings = True  # ignore "Warning: Unknown table ..."
+        cursor = MySQLSilentCursor(self.cnx)
         cursor.execute("DROP TABLE IF EXISTS " + self.quote(table))
 
     def get_column_names(self, table):
